File: zod/src/models/Settlement.ts
import { z } from 'zod';
import { FixedSettlement } from './FixedSettlement';
import { PerHourSettlement } from './PerHourSettlement';

/**
 * 支払い方式を表す列挙型
 */
export const SettlementType = z.enum(['FIXED', 'PER_HOUR']);

/**
 * 支払い方式の型
 */
export type SettlementType = z.infer<typeof SettlementType>;

/**
 * 支払い方式のスキーマ
 */
export const Settlement = z.discriminatedUnion('type', [
  FixedSettlement,
  PerHourSettlement,
]);

/**
 * 支払い方式の型
 */
export type Settlement = z.infer<typeof Settlement>; 

File: yavi/src/main/java/com/example/yavi/joboffer/domain/Settlement.java
package com.example.yavi.joboffer.domain;

import am.ik.yavi.arguments.Arguments1Validator;
import am.ik.yavi.arguments.StringValidator;
import am.ik.yavi.builder.StringValidatorBuilder;

import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.function.Function;

public sealed interface Settlement permits FixedSettlement, PerHourSettlement {
    enum Method { FIXED, PER_HOUR }

    StringValidator<Method> typeValidator = StringValidatorBuilder.of("method",
            c -> c.oneOf(Set.of("fixed", "per_hour"))
    ).build().andThen(s -> Method.valueOf(s.toUpperCase()));

    Arguments1Validator<Map<String, Object>, Settlement> mapValidator = (m, locale, context) ->
            typeValidator.validate(Objects.toString(m.get("type")))
                    .flatMap(method -> (switch (method) {
                        case FIXED -> FixedSettlement.mapValidator.validate(m, locale, context);
                        case PER_HOUR -> PerHourSettlement.mapValidator.validate(m, locale, context);
                    }).map(Function.identity()));
}
